modules/apps/journal/journal-web/src/main/resources/META-INF/resources/js/UndoRedo.js

```diff
@@ -0,0 +1,217 @@
+				`${portletNamespace}${META_FIELD_NAMES.title}`
+			);
+
+			const friendlyURLInputComponent = Liferay.componentReady(
+				`${portletNamespace}${META_FIELD_NAMES.friendlyURL}`
+			);
+
+			const selectedLanguageIdInput = document.getElementById(
+				`${portletNamespace}languageId`
+			);
+
+			Promise.all([
+				descriptionInputComponent,
+				titleInputComponent,
+				friendlyURLInputComponent,
+			]).then(
+				([
+					descriptionInputComponent,
+					titleInputComponent,
+					friendlyURLInputComponent,
+				]) => {
+					const newHistory = {
+						defaultLanguageId: defaultLanguageIdInput.value,
+						descriptionInputComponent:
+							descriptionInputComponent.getValue(
+								selectedLanguageId
+							),
+						friendlyURLInputComponent:
+							friendlyURLInputComponent.getValue(
+								selectedLanguageId
+							),
+						name: fieldName,
+						selectedLanguageId: selectedLanguageIdInput.value,
+						titleInputComponent:
+							titleInputComponent.getValue(selectedLanguageId),
+					};
+
+					setState({
+						defaultLanguageId: defaultLanguageIdInput.value,
+						history: [...history.slice(0, step + 1), newHistory],
+						selectedLanguageId: selectedLanguageIdInput.value,
+						step: step + 1,
+					});
+				}
+			);
+		},
+		[history, portletNamespace, selectedLanguageId, step]
+	);
+
+	const localeChangeHandler = useCallback(
+		(event) => {
+			const selectedLanguageId = event.item.getAttribute('data-value');
+
+			const selectedLanguageIdInput = document.getElementById(
+				`${portletNamespace}languageId`
+			);
+
+			selectedLanguageIdInput.value = selectedLanguageId;
+
+			Liferay.fire('journal:storeState', {fieldName: 'Locale Change'});
+		},
+		[portletNamespace]
+	);
+
+	useEffect(() => {
+		Liferay.after('inputLocalized:localeChanged', localeChangeHandler);
+
+		return () => {
+			Liferay.detach('inputLocalized:localeChanged', localeChangeHandler);
+		};
+	}, [localeChangeHandler]);
+
+	useEffect(() => {
+		Liferay.on('journal:storeState', handleStoreState);
+
+		return () => {
+			Liferay.detach('journal:storeState', handleStoreState);
+		};
+	}, [handleStoreState]);
+
+	useEffect(() => {
+		setTimeout(() => {
+			Liferay.fire('journal:storeState', {fieldName: 'Reset'});
+		}, 1000);
+
+		// eslint-disable-next-line react-hooks/exhaustive-deps
+	}, []);
+
+	return (
+		<>
+			<ClayButtonWithIcon
+				aria-label={Liferay.Language.get('undo')}
+				className="btn-monospaced"
+				disabled={step <= 0}
+				displayType="secondary"
+				onClick={() => {
+					Liferay.fire('journal:undo');
+					handleUndoRedo(step - 1);
+				}}
+				size="sm"
+				symbol="undo"
+				title={Liferay.Language.get('undo')}
+			/>
+
+			<ClayButtonWithIcon
+				aria-label={Liferay.Language.get('redo')}
+				className="btn-monospaced"
+				disabled={!history.length || step === history.length - 1}
+				displayType="secondary"
+				onClick={() => {
+					Liferay.fire('journal:redo');
```

Comment:
These events `Liferay.fire('journal:undo');` and `Liferay.fire('journal:redo');` are not used here (I know they are for DDM). I suggest to remove it from this pull and will add it in the next one, so it is more clear what are they for.